File: LeetCode-Java/Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode start1 = head;
        ListNode start2 = head;
        while(start1 != null && start2 != null && start2.next != null)
        {
            start1 = start1.next;
            start2 = start2.next.next;
            if(start1 == start2)
            {
                return true;
            }
        }
        return false;
    }
}

File: Leetcode-cpp/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *t1 = head;
        ListNode *t2 = head;
        if(head == NULL) return false;
        while(true)
        {
            t1 = t1->next;
            if(t1 == NULL) return false;
            t2 = t2->next;
            if(t2 == NULL) return false;
            t2 = t2->next;
            if(t2 == NULL) return false;
            if(t1 == t2) break;
        }
        return true;
    }
};
